azure-core/azure/core/pipeline/transport/base.py

```diff
@@ -0,0 +1,244 @@
+        """Allow the sender to build a context that will be passed
+        across the pipeline with the request.
+
+        Return type has no constraints. Implementation is not
+        required and None by default.
+        """
+        return None
+
+    def sleep(self, duration):
+        time.sleep(duration)
+
+
+class HttpRequest(object):
+    """Represents a HTTP request.
+
+    URL can be given without query parameters, to be added later using "format_parameters".
+
+    Instance can be created without data, to be added later using "add_content"
+
+    Instance can be created without files, to be added later using "add_formdata"
+
+    :param str method: HTTP method (GET, HEAD, etc.)
+    :param str url: At least complete scheme/host/path
+    :param dict[str,str] headers: HTTP headers
+    :param files: Files list.
+    :param data: Body to be sent.
+    :type data: bytes or str.
+    """
+    def __init__(self, method, url, headers=None, files=None, data=None):
+        # type: (str, str, Mapping[str, str], Any, Any) -> None
+        self.method = method
+        self.url = url
+        self.headers = CaseInsensitiveDict(headers)
+        self.files = files
+        self.data = data
+
+    def __repr__(self):
+        return '<HttpRequest [%s]>' % (self.method)
+
+    @property
+    def body(self):
+        """Alias to data."""
+        return self.data
+
+    @body.setter
+    def body(self, value):
+        self.data = value
+
+    @staticmethod
+    def _format_data(data):
+        # type: (Union[str, IO]) -> Union[Tuple[None, str], Tuple[Optional[str], IO, str]]
+        """Format field data according to whether it is a stream or
+        a string for a form-data request.
+
+        :param data: The request field data.
+        :type data: str or file-like object.
+        """
+        if hasattr(data, 'read'):
+            data = cast(IO, data)
+            data_name = None
+            try:
+                if data.name[0] != '<' and data.name[-1] != '>':
+                    data_name = os.path.basename(data.name)
+            except (AttributeError, TypeError):
+                pass
+            return (data_name, data, "application/octet-stream")
+        return (None, cast(str, data))
+
+    def format_parameters(self, params):
+        # type: (Dict[str, str]) -> None
+        """Format parameters into a valid query string.
+        It's assumed all parameters have already been quoted as
+        valid URL strings.
+
+        :param dict params: A dictionary of parameters.
+        """
+        query = urlparse(self.url).query
+        if query:
+            self.url = self.url.partition('?')[0]
+            existing_params = {
+                p[0]: p[-1]
+                for p in [p.partition('=') for p in query.split('&')]
+            }
+            params.update(existing_params)
+        query_params = ["{}={}".format(k, v) for k, v in params.items()]
+        query = '?' + '&'.join(query_params)
```

Comment:
Not sure if it is a concern but it seems possible that `params` and `existing_params` could both be empty in which case this would add a `?` to `self.url` needlessly.

Reply:
True - it could be cleaned up. Though I don't *think* a redundant '?' makes it an invalid URL.